File: sdk/typescript/src/core/contracts.ts
import { z } from 'zod';

// Core structural boundaries identical to Python's Pydantic `contracts.py`

export const ToolCallSchema = z.object({
    tool_name: z.string(),
    inputs: z.record(z.any()),
    outputs: z.record(z.any()).nullable(),
    error: z.string().nullable().optional(),
    timestamp: z.number(),
});
export type ToolCall = z.infer<typeof ToolCallSchema>;

export const SnapshotSchema = z.object({
    snapshot_hash: z.string(),
    parent_hash: z.string().nullable(),
    data: z.record(z.any()),
    timestamp: z.number(),
});
export type Snapshot = z.infer<typeof SnapshotSchema>;

export const TransactionSchema = z.object({
    transaction_id: z.string().uuid(),
    graph_hash: z.string(),
    model_versions: z.record(z.string()),
    prompt_version: z.string(),
    input_payload: z.record(z.any()),
    memory_snapshot_reference: z.string(),
    tool_call_graph: z.array(ToolCallSchema),
    output_payload: z.record(z.any()).nullable(),
    state_delta: z.record(z.any()).nullable(),
    commit_status: z.enum(['PENDING', 'COMMITTED', 'ROLLED_BACK']),
    timestamp: z.number(),
});
export type Transaction = z.infer<typeof TransactionSchema>;

export const GraphCompiledSchema = z.object({
    graph_name: z.string(),
    version: z.string(),
    nodes: z.array(z.string()),
    edges: z.array(z.tuple([z.string(), z.string()])),
    compiled_hash: z.string(),
    timestamp: z.number()
});
export type GraphCompiled = z.infer<typeof GraphCompiledSchema>;

File: core/contracts.py
import uuid
import time
from enum import Enum
from typing import Dict, Any, List, Optional
from pydantic import BaseModel, Field, ConfigDict

class CommitStatus(str, Enum):
    PENDING = "PENDING"
    COMMITTED = "COMMITTED"
    ROLLED_BACK = "ROLLED_BACK"

class NodeContract(BaseModel):
    node_id: str
    node_type: str  # LLMNode, ToolNode, MemoryNode, ValidationNode
    input_schema: Dict[str, Any]
    output_schema: Dict[str, Any]
    validation_rules: List[str] = Field(default_factory=list)
    guardrails: List[str] = Field(default_factory=list)

class EdgeContract(BaseModel):
    source_id: str
    target_id: str
    data_mapping: Dict[str, str]  # Map source output keys -> target input keys

class Snapshot(BaseModel):
    model_config = ConfigDict(protected_namespaces=())
    
    snapshot_hash: str
    prompt_state: Dict[str, Any]
    model_version: str
    input_tokens: int
    output_tokens: int
    tool_outputs: List[Dict[str, Any]]
    memory_state: Dict[str, Any]
    created_at: float = Field(default_factory=time.time)

class Transaction(BaseModel):
    model_config = ConfigDict(protected_namespaces=())

    transaction_id: str = Field(default_factory=lambda: str(uuid.uuid4()))
    execution_graph_hash: str
    execution_graph_version: str = "1.0.0"
    model_versions: Dict[str, str]
    prompt_version: str
    input_payload: Dict[str, Any]
    memory_snapshot_reference: str
    tool_call_graph: List[Dict[str, Any]] = Field(default_factory=list)
    output_payload: Optional[Dict[str, Any]] = None
    state_delta: Optional[Dict[str, Any]] = None
    commit_status: CommitStatus = CommitStatus.PENDING
    timestamp: float = Field(default_factory=time.time)
